### ISSUE TITLE
DKMS build error on 4.13

### ISSUE THREAD

**Issue by monotykamary** [dtaht/sch_cake#62]
This is a patched kernel, however the patches were previously applied on 4.12 with no problems to the dkms build.

```
DKMS make.log for sch_cake-r266.e641a56 for kernel 4.13.0-2-bfq-mq (x86_64)
Tue Sep  5 02:18:20 +07 2017
make: Entering directory '/usr/lib/modules/4.13.0-2-bfq-mq/build'
  AR      /var/lib/dkms/sch_cake/r266.e641a56/build/built-in.o
  CC [M]  /var/lib/dkms/sch_cake/r266.e641a56/build/sch_cake.o
/var/lib/dkms/sch_cake/r266.e641a56/build/sch_cake.c:2074:3: error: ‘const struct Qdisc_class_ops’ has no member named ‘tcf_chain’
  .tcf_chain = cake_find_tcf,
   ^~~~~~~~~
/var/lib/dkms/sch_cake/r266.e641a56/build/sch_cake.c:2074:15: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .tcf_chain = cake_find_tcf,
               ^~~~~~~~~~~~~
/var/lib/dkms/sch_cake/r266.e641a56/build/sch_cake.c:2074:15: note: (near initialization for ‘cake_class_ops.change’)
cc1: some warnings being treated as errors
make[1]: *** [scripts/Makefile.build:309: /var/lib/dkms/sch_cake/r266.e641a56/build/sch_cake.o] Error 1
make: *** [Makefile:1512: _module_/var/lib/dkms/sch_cake/r266.e641a56/build] Error 2
make: Leaving directory '/usr/lib/modules/4.13.0-2-bfq-mq/build'
```

**Comment by koenkooi** [comment]
I can confirm the same issue with an unpatched 4.13.1 kernel

**Comment by Lochnair** [comment]
Relevant kernel change: https://github.com/torvalds/linux/commit/6529eaba33f0465fc6d228e1d05b1745f7d0e8c9

**Comment by dtaht** [comment]
Nils Andreas Svee <notifications@github.com> writes:

> Relevant kernel change: torvalds/linux@6529eab

Sigh. It seemed like the simplest possible patch was to merely change
the .tcf_chain to .tcf_block (as cake doesn't support tc filters),
but my very quick attempt at then compiling that against net-next blew up with:

In file included from /home/dave/git/sch_cake/sch_cake.c:57:0:
./include/net/pkt_sched.h: In function ‘is_classid_clsact_ingress’:
./include/net/pkt_sched.h:139:39: error: ‘TC_H_CLSACT’ undeclared (first use in this function)
  return TC_H_MAJ(classid) == TC_H_MAJ(TC_H_CLSACT) &&
                                       ^
/home/dave/git/sch_cake/pkt_sched.h:68:23: note: in definition of macro ‘TC_H_MAJ’
 #define TC_H_MAJ(h) ((h)&TC_H_MAJ_MASK)
                       ^
./include/net/pkt_sched.h:139:39: note: each undeclared identifier is reported only once for each function it appears in
  return TC_H_MAJ(classid) == TC_H_MAJ(TC_H_CLSACT) &&
                                       ^
/home/dave/git/sch_cake/pkt_sched.h:68:23: note: in definition of macro ‘TC_H_MAJ’
 #define TC_H_MAJ(h) ((h)&TC_H_MAJ_MASK)
                       ^
./include/net/pkt_sched.h:140:39: error: ‘TC_H_MIN_EGRESS’ undeclared (first use in this function)
         TC_H_MIN(classid) != TC_H_MIN(TC_H_MIN_EGRESS);
                                       ^
/home/dave/git/sch_cake/pkt_sched.h:69:23: note: in definition of macro ‘TC_H_MIN’
 #define TC_H_MIN(h) ((h)&TC_H_MIN_MASK)
                       ^
./include/net/pkt_sched.h: In function ‘is_classid_clsact_egress’:
./include/net/pkt_sched.h:145:39: error: ‘TC_H_CLSACT’ undeclared (first use in this function)
  return TC_H_MAJ(classid) == TC_H_MAJ(TC_H_CLSACT) &&
                                       ^
/home/dave/git/sch_cake/pkt_sched.h:68:23: note: in definition of macro ‘TC_H_MAJ’
 #define TC_H_MAJ(h) ((h)&TC_H_MAJ_MASK)
                       ^
./include/net/pkt_sched.h:146:39: error: ‘TC_H_MIN_EGRESS’ undeclared (first use in this function)
         TC_H_MIN(classid) == TC_H_MIN(TC_H_MIN_EGRESS);
                                       ^
/home/dave/git/sch_cake/pkt_sched.h:69:23: note: in definition of macro ‘TC_H_MIN’
 #define TC_H_MIN(h) ((h)&TC_H_MIN_MASK)
                       ^
/home/dave/git/sch_cake/sch_cake.c: In function ‘cake_change’:
/home/dave/git/sch_cake/sch_cake.c:1688:8: error: too few arguments to function ‘nla_parse_nested’
  err = nla_parse_nested(tb, TCA_CAKE_MAX, opt, cake_policy);
        ^
In file included from /home/dave/git/sch_cake/sch_cake.c:55:0:
./include/net/netlink.h:759:19: note: declared here
 static inline int nla_parse_nested(struct nlattr *tb[], int maxtype,
                   ^
/home/dave/git/sch_cake/sch_cake.c: At top level:
/home/dave/git/sch_cake/sch_cake.c:2068:2: error: unknown field ‘get’ specified in initializer
  .get  = cake_get,
  ^
/home/dave/git/sch_cake/sch_cake.c:2068:10: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .get  = cake_get,
          ^
/home/dave/git/sch_cake/sch_cake.c:2068:10: note: (near initialization for ‘cake_class_ops.qlen_notify’)
/home/dave/git/sch_cake/sch_cake.c:2069:2: error: unknown field ‘put’ specified in initializer
  .put  = cake_put,
  ^
/home/dave/git/sch_cake/sch_cake.c:2069:10: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .put  = cake_put,
          ^
/home/dave/git/sch_cake/sch_cake.c:2069:10: note: (near initialization for ‘cake_class_ops.find’)
/home/dave/git/sch_cake/sch_cake.c:2070:15: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .tcf_block = cake_find_tcf,
               ^
/home/dave/git/sch_cake/sch_cake.c:2070:15: note: (near initialization for ‘cake_class_ops.tcf_block’)
cc1: some warnings being treated as errors
scripts/Makefile.build:317: recipe for target '/home/dave/git/sch_cake/sch_cake.o' failed
make[2]: *** [/home/dave/git/sch_cake/sch_cake.o] Error 1
Makefile:1512: recipe for target '_module_/home/dave/git/sch_cake' failed
make[1]: *** [_module_/home/dave/git/sch_cake] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.13.0-netem-1'
Makefile:8: recipe for target 'default' failed
make: *** [default] Error 2
dave@nemesis:~/git/sch_cake$

**Comment by chromi** [comment]
Having finally unearthed my work to date from a damaged computer, I think
I've also managed to solve this issue by simply deleting the sub "class
handling" code left over from the original use of fq_codel as a template.
In my testing, the cobalt branch now builds and runs successfully against
both 4.13 and 4.11 kernels.

Today's update also addresses the outstanding objection to ingress mode.
So perhaps a promotion of that to the master branch is also practical soon.

- Jonathan Morton

**Comment by dtaht** [comment]
chromi <notifications@github.com> writes:

> Having finally unearthed my work to date from a damaged computer, I think
> I've also managed to solve this issue by simply deleting the sub "class
> handling" code left over from the original use of fq_codel as a template.
> In my testing, the cobalt branch now builds and runs successfully against
> both 4.13 and 4.11 kernels.
>
> Today's update also addresses the outstanding objection to ingress mode.
> So perhaps a promotion of that to the master branch is also practical soon.

compiling from net-next suffered from some header movement

so this wrapped with some suitable version macro fixes that.

diff --git a/pkt_sched.h b/pkt_sched.h
index bdbf07d..3027cb1 100644
--- a/pkt_sched.h
+++ b/pkt_sched.h
@@ -63,6 +63,7 @@ struct tc_estimator {
     Macros to manipulate handles:
  */
 
+
 #define TC_H_MAJ_MASK (0xFFFF0000U)
 #define TC_H_MIN_MASK (0x0000FFFFU)
 #define TC_H_MAJ(h) ((h)&TC_H_MAJ_MASK)
@@ -72,6 +73,10 @@ struct tc_estimator {
 #define TC_H_UNSPEC	(0U)
 #define TC_H_ROOT	(0xFFFFFFFFU)
 #define TC_H_INGRESS    (0xFFFFFFF1U)
+#define TC_H_CLSACT	TC_H_INGRESS
+
+#define TC_H_MIN_INGRESS	0xFFF2U
+#define TC_H_MIN_EGRESS		0xFFF3U
 
 /* Need to corrospond to iproute2 tc/tc_core.h "enum link_layer" */
 enum tc_link_layer {

**Comment by dtaht** [comment]
I just built the cobalt branch. Very practical to eliminate all tc
support... :)

I have no real idea what the differences are between the cobalt
branch and the main branch at this point. The diff was unenlightening
for the most part.

How should one go about testing ingress mode?

**Comment by chromi** [comment]
On 15/09/17 02:59, Dave Täht wrote:
> I have no real idea what the differences are between the cobalt
> branch and the main branch at this point. The diff was unenlightening
> for the most part.

The most significant new feature is ingress mode, of course.  The 
required userspace stuff for that is already in tc-adv.  It may still 
require some revisions to the documentation.

I also rewrote the tin choice logic in an attempt to improve latency 
control.  This doesn't seem to have entirely succeeded in practice, but 
the new logic is theoretically cleaner, using the same "virtual clock" 
basis as the shaper, and is certainly no worse.  Associated with that is 
a fix for the way diffserv-llt works.

And now we have kernel 4.13 compatibility, by deleting some unused code. 
  If and when we get around to writing a proper packet classification 
engine, we should examine whether making that the parent or child of a 
Cake instance is the best choice, and add classfulness support 
appropriately.

The above would probably be easier to discern if I had been able to 
preserve my full Git history.  Circumstances conspired against that.

> How should one go about testing ingress mode?

The change I've made to ingress mode ensures that *some* traffic always 
gets through even if the drop rate exceeds the shaped rate.  This is 
done by running a second shaper at a reduced rate, without the special 
treatment of dropped packets, in parallel with the main one.

The test I ran to prove that involved stressing a low-bandwidth shaper 
with an unresponsive flood of traffic while trying to get some real 
traffic through it.  Variations on that theme might be interesting.

When all the traffic is either sparse or ECN-enabled and responsive 
(including arbitrary mixtures of those types), the drop rate should 
remain zero and full throughput should be maintained.  That, I think, 
would be worth verifying.

  - Jonathan Morton
